core/arch/x86/decode.c

```diff
@@ -1446,20 +1456,24 @@ decode_reg(decode_reg_t which_reg, decode_info_t *di, byte optype, opnd_size_t o
     case TYPE_VSIB: {
         reg_id_t extend_reg = extend ? reg + 8 : reg;
         extend_reg = avx512_extend ? extend_reg + 16 : extend_reg;
-        return (TEST(PREFIX_EVEX_LL, di->prefixes)
-                    ? (DR_REG_START_ZMM + extend_reg)
-                    : ((TEST(PREFIX_VEX_L, di->prefixes) &&
-                        /* Not only do we use this for VEX .LIG (where raw reg is either
-                         * OPSZ_32 or OPSZ_16_vex32) but also for VSIB which currently
-                         * does not get up to OPSZ_16 so we can use this negative
-                         * check.
-                         * XXX i#1312: vgather/vscatter VSIB addressing may be OPSZ_16?
-                         * For EVEX .LIG, raw reg will be able to be OPSZ_64 or
-                         * OPSZ_16_vex32_evex64.
-                         */
-                        expand_subreg_size(opsize) != OPSZ_16)
-                           ? (REG_START_YMM + extend_reg)
-                           : (REG_START_XMM + extend_reg)));
+        bool operand_is_zmm = TEST(PREFIX_EVEX_LL, di->prefixes) &&
+            expand_subreg_size(opsize) != OPSZ_16 &&
+            expand_subreg_size(opsize) != OPSZ_32;
+        /* Not only do we use this for VEX .LIG and EVEX .LIG (where raw reg is
+         * either OPSZ_32 or OPSZ_16_vex32 or OPSZ_vex32_evex64) but also
+         * for VSIB which currently does not get up to OPSZ_16 so we can
+         * use this negative check.
+         * XXX i#1312: vgather/vscatter VSIB addressing may be OPSZ_16?
+         * For EVEX .LIG, raw reg will be able to be OPSZ_64 or
+         * OPSZ_16_vex32_evex64.
+         */
+        bool operand_is_ymm = (TEST(PREFIX_EVEX_LL, di->prefixes) &&
+                               expand_subreg_size(opsize) == OPSZ_32) ||
+            (TEST(PREFIX_VEX_L, di->prefixes) && expand_subreg_size(opsize) != OPSZ_16);
```

Comment:
Not sure this !=16 makes sense either: what if it's 8 or sthg else.  Here there is a mention of the negative check in the comment about VSIB, but there are other sizes returned now in expand_subreg_size() so it may need re-evaluation.

Reply:
Same comment as above. I agree.